### ISSUE TITLE
[BUG] Metal 2-pass vector SDPA kernel produces garbage decode output on M3 Max (applegpu_g15s) with GQA once KV length reaches 1024

### ISSUE THREAD

**Issue by beyondsimulations** [ml-explore/mlx#4022]
### Describe the bug

On an M3 Max (`applegpu_g15s`), single-token decode routed to
`sdpa_vector_2pass` produces numerically wrong output for a GQA model
(32 query heads / 8 KV heads, head_dim 128). Generation is correct for every
step routed to the 1-pass `sdpa_vector` kernel and collapses into a runaway of
`!` tokens starting at exactly the first step whose KV length reaches the
2-pass routing threshold of 1024.

The dispatch site is `mlx/backend/metal/scaled_dot_product_attention.cpp`
(lines 740–750 at `ce45c52`, tag `v0.31.1`):

```cpp
// We route to the 2 pass fused attention if
// - The device is large and the sequence length long
// - The sequence length is even longer and we have gqa
bool do_causal = do_causal_ && q.shape(2) > 1;
char devc = d.get_architecture().back();
if (((devc == 'd' || devc == 's') && k.shape(2) >= 1024) ||
    (k.shape(1) < q.shape(1) && k.shape(2) >= 4096)) {
  sdpa_vector_2pass(s, d, q, k, v, o, scale_, do_causal, mask, sinks);
} else {
  sdpa_vector(s, d, q, k, v, o, scale_, do_causal, mask, sinks);
}
```

`applegpu_g15s` ends in `'s'`, so the first condition fires at `k.shape(2) == 1024`.

I have **not** isolated the fault inside the 2-pass kernel itself. What I can
show is that the failure is exactly co-located with this dispatch decision:
forcing every vector-path decode onto `sdpa_vector` makes the output correct,
with nothing else changed.

For this configuration `sdpa_vector_2pass` selects `blocks = 64` (device char
`'s'`, `n_simds = gqa_factor * q.shape(2) = 4 * 1 = 4`, so the `n_simds > 4`
escalation does not apply), `BN = 32`, `head_dim = 128`.

### To reproduce

I do not have a standalone pure-MLX numeric repro — the failure was found and
pinned in a Swift VLM generation stack. I can build a minimal one on request if
that would help; the setup below is 100% deterministic and reproduces every run.

1. Stack: `mlx-swift` (Cmlx core pinned at `ce45c52` = `v0.31.1`) +
   `ml-explore/mlx-swift-lm` at `3cbf928b5eb24190e8952725699ae6a3bb02824d`,
   using `MLXVLM` / `MLXLMCommon`.
2. Model: `mlx-community/Qwen3-VL-8B-Instruct-4bit` (affine, 4 bits, group 64).
   Text config: 36 layers, `num_attention_heads: 32`, `num_key_value_heads: 8`,
   `head_dim: 128`, `hidden_size: 4096`.
3. Single image + short instruction prompt, sized so that vision tokens +
   prompt land a few hundred tokens *below* 1024 (mine: a 1404x1872 page image
   with the pixel budget clamped to 786,432, giving a prompt KV of 905).
4. `GenerateParameters(maxTokens: 1024, temperature: 0)`. Default KV cache —
   `KVCacheSimple`, `maxKVSize` unset, so no rotation and no cache-side
   quantization. Nothing in my configuration is parameterized at 1024.
5. Generate.

With an instrumented build printing `k.shape(2)`, `q.shape(1)`, `k.shape(1)`
and the branch taken at the dispatch site, the run produces:

```
[SDPA_ROUTE] kv=1022 qh=32 kvh=8 -> 1pass
[SDPA_ROUTE] kv=1023 qh=32 kvh=8 -> 1pass
[SDPA_ROUTE] kv=1024 qh=32 kvh=8 -> 2pass
[SDPA_ROUTE] kv=1024 qh=32 kvh=8 -> 2pass
```

with 4,284 `1pass` calls and 32,580 `2pass` calls. 4,284 / 36 layers = 119
decode steps, and 905 + 119 = 1024 — i.e. every step before the routing flip is
1-pass, every step from the flip onward is 2-pass.

The generated text has exactly 119 sensible tokens followed by 904 `!` tokens
(1,023 total against `maxTokens: 1024`), so the collapse begins on the very
first token computed by the 2-pass kernel.

The collapse point tracks KV length, not content: with a different image /
prompt length the clean prefix grows or shrinks so that the break still lands
at cumulative KV 1024.

### Expected behavior

`sdpa_vector_2pass` and `sdpa_vector` produce the same output (within fp
tolerance), so crossing the routing threshold mid-generation is invisible.

### Actual behavior

Output is correct up to the routing flip and then degenerates immediately into
a `!`/whitespace runaway that continues until the token budget is exhausted.
Sanitized shape of the transcript (content replaced, structure preserved):

```
- [ ] first item, transcribed correctly
- [ ] second item, transcribed correctly
...
- [ ] Check if Ju(!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
  (904 '!' tokens, to the token limit)
```

It cuts in mid-word, at temperature 0 — this is not a sampling artifact, and
`!` is the token the logits collapse onto.

### Environment

- macOS 26.5.1 (build 25F80)
- MacBook Pro, `Mac15,9`, Apple M3 Max, 16-core CPU, 48 GB unified memory
- Metal GPU architecture: `applegpu_g15s` (so `devc == 's'`)
- mlx core `v0.31.1` (`ce45c52`), consumed as the `Cmlx` submodule of
  `ml-explore/mlx-swift`
- `ml-explore/mlx-swift-lm` @ `3cbf928b5eb24190e8952725699ae6a3bb02824d`
- Model: `mlx-community/Qwen3-VL-8B-Instruct-4bit`

### Additional context

**Workaround (2-line dispatch patch).** Raising both routing thresholds out of
reach forces the vector path onto the 1-pass kernel and fixes the output
completely:

```diff
@@ -742,8 +742,8 @@ void ScaledDotProductAttention::eval_gpu(
-    if (((devc == 'd' || devc == 's') && k.shape(2) >= 1024) ||
-        (k.shape(1) < q.shape(1) && k.shape(2) >= 4096)) {
+    if (((devc == 'd' || devc == 's') && k.shape(2) >= 100000000) ||
+        (k.shape(1) < q.shape(1) && k.shape(2) >= 100000000)) {
       sdpa_vector_2pass(s, d, q, k, v, o, scale_, do_causal, mask, sinks);
     } else {
       sdpa_vector(s, d, q, k, v, o, scale_, do_causal, mask, sinks);
```

Branch: https://github.com/byndsim/mlx/tree/markink/sdpa-1pass
(commit `d8964f06ae1d4b0f856bfb4d4b4a9437cf9b77ed`, based on `ce45c52`).

With that patch, the same 22 inputs run clean over a long-KV batch, verified to
a maximum KV length of 6,022 with zero 2-pass dispatches and no `!` runaway on
any page. This is obviously not a fix — it forfeits the 2-pass kernel's
long-context performance — but it does localize the fault to the 2-pass path.

**Not fixed in v0.32.0.** I diffed `v0.31.1..v0.32.0`:

- `mlx/backend/metal/kernels/sdpa_vector.h` is byte-identical (empty diff), so
  both 2-pass kernel bodies are unchanged.
- `scaled_dot_product_attention.cpp` changed (15 insertions / 11 deletions) but
  the two routing thresholds are untouched; the changes are the
  `metal::get_command_encoder(s)` refactor, a 192/128 head-dim allowance in
  `use_fallback`, and a new `MLX_SDPA_BLOCKS` env override in
  `sdpa_vector_2pass`. The `blocks` selection logic itself is unchanged.

I have not tested v0.32.0 end-to-end (the version bump does not build against
the mlx-c revision my Swift stack pins), so this is a source-level rather than
a runtime finding.

The new `MLX_SDPA_BLOCKS` knob in 0.32.0 may make this easy to bisect from
Python — if the failure turns out to be block-count-dependent, that would also
explain why Python users on the same hardware might not have hit it (I have not
been able to confirm either way; I have no Python mlx install on this machine).

**Still present on current `main` (source-level, checked 2026-08-05):** the
2-pass routing thresholds and the `blocks` selection heuristic in
`scaled_dot_product_attention.cpp` are structurally unchanged from `v0.31.1`,
and `sdpa_vector.h`'s 2-pass kernel bodies likewise. I see #4018 is currently
touching the 1-pass `sdpa_vector` dispatch (threadgroup-size check), so this
area is active — happy to rebase my instrumentation onto whatever lands there.
Possibly-related numerics reports I found while searching (neither matches this
failure shape): #3585 (fused-vs-manual float ordering, with a Metal-toolchain
theory) and #3856 (M5 quantized-MoE corruption).

**Offer:** I ship this workaround in a production app and maintain a
deterministic RED/GREEN regression test for it (a whole-page VLM transcription
that runs KV well past 1024 and asserts a clean, non-degenerate result — RED on
the stock kernel, GREEN with the dispatch patch, runs in under a minute with
weights installed). I'm glad to run that gate against candidate fixes on the
affected M3-class hardware, provide the full instrumented routing log and exact
prompt/image setup, or build a minimal standalone repro against Python mlx —
just say which is most useful.

**Comment by katlun-lgtm** [comment]
I have the same GPU (`applegpu_g15s`, M3 Max) so I built the standalone pure-MLX repro you offered, and **I can't make it fail**. Comparing `mx.fast.scaled_dot_product_attention` against a float32 reference at your shape (32 q heads / 8 KV heads, head_dim 128, qL=1), 3 seeds per cell:

| axis | covered | worst rel err |
|---|---|---|
| version | **0.31.1** and 0.32.0 | identical results |
| kv_len | 512, 1000, 1023, **1024**, 1025, 1100, 2048, 4096, 8192 | no discontinuity at 1024 |
| dtype | bf16 / fp16 | 4.1e-3 / 4.6e-4 |
| K,V form | contiguous, and **sliced out of a preallocated 16384 cache** | 3.4e-3 |
| mask | none, all-zero additive, **−inf prefix masking 25–98% of keys** | 3.8e-3 |

No NaN anywhere; everything is bf16/fp16 rounding. Worth noting 0.31.1 and 0.32.0 give the same numbers, so this isn't something that got fixed after your version.

Two things I deliberately tried because a naive repro wouldn't: passing K/V as a **slice of a larger preallocated buffer** (what a real KV cache hands you, so any stride/offset assumption would show up), and **fully-masked blocks** via −inf prefixes — a block whose partial max is −inf and partial sum is 0 is the classic NaN source in a 2-pass softmax combine, and it would explain why your failure starts exactly when block splitting starts. Both are clean.

So I don't think the fault is the 2-pass kernel itself at this shape, which probably means something your Swift stack passes that I haven't modelled. Could you say whether any of these are in play: quantized KV cache, attention sinks, steps where qL > 1, a head_dim other than 128, or a non-contiguous query? Happy to post the scripts if you want to extend them on your side.

**Comment by beyondsimulations** [comment]
Thanks for taking the time to build this!!

To your questions: no quantized KV cache, no sinks, head_dim is 128 in the decode loop, the query is contiguous, and every 2-pass call is a single-token decode step. Nothing exotic on my side that I can find.

Small complication: I'm on vacation and only have a MacBook Air (M5) with me, not the M3 Max where it fails. On the Air the 2-pass kernel kicks in at kv 4096 instead of 1024, so I tested what I could there. Short version: I can't break it either, even my full Swift app
rebuilt against stock mlx-swift with no patch. All clean on the Air.

So whatever this is, it seems to need the M3 Max specifically. I'll be back at that machine in about two weeks and will run two things there: the same real-model generation through the Python wheel (if it breaks, there's your standalone repro), and if that's clean, my Swift build with the same inputs, which would leave the locally compiled Metal kernels as the main suspect. Related: which macOS/Xcode did you build with?

Please do post your scripts, I'll run them on both machines when I'm back.

**Comment by katlun-lgtm** [comment]
Good to rule those out. Build env here is macOS 27.0 (26A5388g) / Xcode 26.6 (17F113), Metal toolchain whatever ships with that — let me know if you need something more specific once you can check the Swift side.

Scripts are here: https://gist.github.com/katlun-lgtm/53e4fe90b419f4b6814b0b48a8aa3221 (the three from round 1/2/3 above). No rush, whenever you're back at the M3 Max.

**Comment by zcbenz** [comment]
Closing due to lack of activity.
